src/java.base/share/classes/sun/security/ssl/ServerHello.java

```diff
@@ -794,6 +794,15 @@ public byte[] produce(ConnectionContext context,
             hhrm.write(shc.handshakeOutput);
             shc.handshakeOutput.flush();
 
+            // In the Middlebox Compatibility Mode the server sends a
+            // dummy change_cipher_spec record immediately after its
+            // first handshake message. This may either be after
+            // a ServerHello or a HelloRetryRequest.
+            //(RFC 8446, Appendix D.4)
+            shc.conContext.outputRecord.changeWriteCiphers(
+                SSLWriteCipher.nullTlsWriteCipher(),
+                    (clientHello.sessionId.length() != 0));
```

Comment:
```suggestion
            shc.conContext.outputRecord.encodeChangeCipherSpec();
```

Reply:
Unfortunately, we lack separate properties to control named groups in both the server and client. When running server and client threads in the same JVM, manipulating client hello packets to prompt the server to trigger HRR becomes exceedingly challenging.

Reply:
since [JDK-8281236](https://bugs.openjdk.org/browse/JDK-8281236) / 5d4c71c8bd361af78c90777f17b79e95d8eb5afe / JDK 20 we have setNamedGroups function to control named groups on every endpoint.

Reply:
Thank you, I added a test

Reply:
> 

IMO, utilizing shc.conContext.outputRecord.changeWriteCiphers() is a cleaner approach, and we've used the same method in other instances.

Reply:
Thanks for adding the test.

My main concern with using changeWriteCiphers here is that it sends the wrong message to the future readers of this code. It suggests that we want to actually change the cipher, and sending CCS is just a side effect. Note that all other uses of changeWriteCiphers actually change the write ciphers. If you don't want to use the other method, at least add a comment explaining why you're using it here.

Reply:
Thanks for the review, in the comments I mentioned that,  this call sends a dummy change_cipher_spec (CCS) record. I hope, It explains why we are calling it here.

Reply:
Will it produce 2 ChangeCipherSpec record. One after HRR and other after SH?

Reply:
Yes, the server produces 2 CCS records in the case of HRR. According to RFC:

"Either side can send change_cipher_spec at any time during the handshake, as they must be ignored by the peer, but if the client sends a non-empty session ID, the server MUST send the change_cipher_spec as described in this
appendix."

https://datatracker.ietf.org/doc/html/rfc8446#appendix-D.4

Reply:
I am not an expert in this field and expressing one of my thought here and my assumption could be wrong too.
Shouldn't it check "SSLConfiguration.useCompatibilityMode" for any change applicable to solve middlebox issue?

Reply:
(clientHello.sessionId.length() != 0) condition checks for same

Reply:
Got it. Thanks..

Reply:
just see this comment regards `useCompatibilityMode` - I'd a similar concern. shouldn't useCompatibilityMode be checked no matter what value we get for `clientHello.sessionId.length() `?

Reply:
`useCompatibilityMode` is a client-side setting. See [the spec](https://www.rfc-editor.org/rfc/rfc8446#page-141):
> if the client sends a non-empty session ID, the server MUST send the change_cipher_spec as described in this appendix.

Reply:
thanks - I'd missed that this is used on client side only

Reply:
Just for completeness, QUIC is using a non-null cipher for client/server hello, so setting a null cipher here would break it. But then, QUIC forbids middlebox compatibility mode, so this shouldn't be an issue.